execution_engine/src/shared/gas.rs

```diff
@@ -40,170 +42,195 @@ impl fmt::Display for Gas {
     }
 }
 
-impl std::ops::Add for Gas {
-    type Output = Gas;
+impl CheckedAdd for Gas {
+    fn checked_add(&self, v: &Self) -> Option<Self> {
+        self.0.checked_add(v.0).map(Gas)
+    }
+}
 
-    fn add(self, rhs: Self) -> Self::Output {
-        let val = self.value() + rhs.value();
-        Gas::new(val)
+impl CheckedMul for Gas {
+    fn checked_mul(&self, v: &Self) -> Option<Self> {
+        self.0.checked_mul(v.0).map(Gas)
     }
 }
 
-impl std::ops::Sub for Gas {
-    type Output = Gas;
+impl CheckedSub for Gas {
+    fn checked_sub(&self, v: &Self) -> Option<Self> {
+        self.0.checked_sub(v.0).map(Gas)
+    }
+}
 
-    fn sub(self, rhs: Self) -> Self::Output {
-        let val = self.value() - rhs.value();
-        Gas::new(val)
+impl Sum for Gas {
+    fn sum<I: Iterator<Item = Self>>(iter: I) -> Self {
+        iter.fold(Gas::zero(), std::ops::Add::add)
     }
 }
 
-impl std::ops::Div for Gas {
-    type Output = Gas;
+impl Integer for Gas {
+    fn div_floor(&self, other: &Self) -> Self {
+        Gas(self.0.div_floor(&other.0))
+    }
 
-    fn div(self, rhs: Self) -> Self::Output {
-        let val = self.value() / rhs.value();
-        Gas::new(val)
+    fn mod_floor(&self, other: &Self) -> Self {
+        Gas(self.0.mod_floor(&other.0))
     }
-}
 
-impl std::ops::Mul for Gas {
-    type Output = Gas;
+    fn gcd(&self, other: &Self) -> Self {
+        Gas(self.0.gcd(&other.0))
+    }
 
-    fn mul(self, rhs: Self) -> Self::Output {
-        let val = self.value() * rhs.value();
-        Gas::new(val)
+    fn lcm(&self, other: &Self) -> Self {
+        Gas(self.0.lcm(&other.0))
     }
-}
 
-impl std::ops::AddAssign for Gas {
-    fn add_assign(&mut self, rhs: Self) {
-        self.0 += rhs.0
+    fn divides(&self, other: &Self) -> bool {
+        self.0.divides(&other.0)
     }
-}
 
-impl Zero for Gas {
-    fn zero() -> Self {
-        Gas::new(U512::zero())
+    fn is_multiple_of(&self, other: &Self) -> bool {
+        self.0.is_multiple_of(&other.0)
     }
 
-    fn is_zero(&self) -> bool {
-        self.0.is_zero()
+    fn is_even(&self) -> bool {
+        self.0.is_even()
     }
-}
 
-impl Sum for Gas {
-    fn sum<I: Iterator<Item = Self>>(iter: I) -> Self {
-        iter.fold(Gas::zero(), std::ops::Add::add)
+    fn is_odd(&self) -> bool {
+        self.0.is_odd()
+    }
+
+    fn div_rem(&self, other: &Self) -> (Self, Self) {
+        let (div, rem) = self.0.div_rem(&other.0);
+        (Gas(div), Gas(rem))
     }
 }
 
 impl From<u32> for Gas {
     fn from(gas: u32) -> Self {
-        let gas_u512: U512 = gas.into();
-        Gas::new(gas_u512)
+        Gas::new(gas as u64)
     }
 }
 
 impl From<u64> for Gas {
     fn from(gas: u64) -> Self {
-        let gas_u512: U512 = gas.into();
-        Gas::new(gas_u512)
+        Gas::new(gas)
+    }
+}
+
+impl From<usize> for Gas {
```

Comment:
Like I mentioned [above][1], when do we cast a `usize` into `Gas`?  

[1]: https://github.com/casper-network/casper-node/pull/1968/files#r700622879

Reply:
https://github.com/casper-network/casper-node/blob/2e08ea52fa54fcabe3c3c9b3badd655a7dc9f350/execution_engine/src/shared/storage_costs.rs#L31 whenever we charge for storing arbitrary number of bytes